cabal-install/bootstrap.py

```diff
@@ -0,0 +1,299 @@
+
+ghc_path = Path('ghc')
+ghc_pkg_path = Path('ghc-pkg')
+
+PACKAGES = Path('packages')
+PKG_DB = PACKAGES / 'packages.conf'
+BOOTSTRAP_DEPS_JSON = Path('bootstrap-deps.json')
+
+PackageName = NewType('PackageName', str)
+Version = NewType('Version', str)
+
+class PackageSource(Enum):
+    HACKAGE = 'hackage'
+    PREEXISTING = 'pre-existing'
+    LOCAL = 'local'
+
+BootstrapDep = NamedTuple('BootstrapDep', [
+    ('package', PackageName),
+    ('version', Version),
+    ('source', PackageSource),
+    # only valid when source == HACKAGE
+    ('revision', Optional[int]),
+    ('sha256', Optional[bytes]),
+])
+
+class BadTarball(Exception):
+    def __init__(self, path: str, expected_sha256: bytes, found_sha256: bytes):
+        self.path = path
+        self.expected_sha256 = expected_sha256
+        self.found_sha256 = found_sha256
+
+    def __str__(self):
+        return '\n'.join([
+            f'Bad tarball hash: {self.path}',
+            f'  expected: {self.expected_sha256}',
+            f'  found:    {self.found_sha256}',
+        ])
+
+def package_url(package: PackageName, version: Version) -> str:
+    return f'https://hackage.haskell.org/package/{package}-{version}/{package}-{version}.tar.gz'
+
+def package_cabal_url(package: PackageName, version: Version, revision: int) -> str:
+    return f'https://hackage.haskell.org/package/{package}-{version}/revision/{revision}.cabal'
+
+def fetch_package(package: PackageName,
+                  version: Version,
+                  revision: Optional[int],
+                  sha256: bytes
+                  ) -> Path:
+    import urllib.request
+
+    # Download source distribution
+    out = PACKAGES / (f'{package}-{version}.tar.gz')
+    if not out.exists():
+        print(f'Fetching {package}-{version}...')
+        out.parent.mkdir(parents=True, exist_ok=True)
+        url = package_url(package, version)
+        with urllib.request.urlopen(url) as resp:
+            shutil.copyfileobj(resp, out.open('wb'))
+
+    # Download revised cabal file
+    cabal_file = PACKAGES / f'{package}.cabal'
+    if revision is not None and not cabal_file.exists():
+        url = package_cabal_url(package, version, revision)
+        with urllib.request.urlopen(url) as resp:
+            shutil.copyfileobj(resp, cabal_file.open('wb'))
+
+    h = hash_file(hashlib.sha256(), out.open('rb'))
+    if sha256 != h:
+        raise BadTarball(out, sha256, h)
+
+    return out
+
+def read_bootstrap_deps(path: Path) -> List[BootstrapDep]:
+    deps = json.load(path.open())
+    def from_json(o: object) -> BootstrapDep:
+        o['source'] = PackageSource(o['source'])
+        return BootstrapDep(**o)
+
+    return [from_json(dep) for dep in deps]
+
+def write_bootstrap_deps(deps: List[BootstrapDep]):
+    def to_json(dep: BootstrapDep) -> object:
+        return {
+            'package': dep.package,
+            'version': dep.version,
+            'source': dep.source.value,
+            'revision': dep.revision,
```

Comment:
revision file should have sha256 too.

Reply:
Fair enough. However, I am going to hold off on addressing this until we have a scheme for figuring out the current revision of a package.

Reply:
As said below, cleanest approach would be to able directly ask `hackage-server` for `package.cabal` with known sha256.